cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,488 @@
+
+    https://gitlab.com/JoD/exact
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_exact
+"""
+import sys  # for stdout checking
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import *
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar
+from ..transformations.comparison import only_numexpr_equality
+from ..transformations.flatten_model import flatten_constraint, flatten_objective, get_or_make_var
+from ..transformations.get_variables import get_variables
+from ..transformations.linearize import linearize_constraint, only_positive_bv, only_const_rhs, only_var_lhs
+from ..transformations.reification import only_bv_implies, reify_rewrite
+import numpy as np
+
+class CPM_exact(SolverInterface):
+    """
+    Interface to the Python interface of Exact
+
+    Requires that the 'exact' python package is installed:
+    $ pip install exact
+    See https://pypi.org/project/exact for more information.
+
+    Creates the following attributes (see parent constructor for more):
+    xct_solver: the Exact instance used in solve()
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import exact
+            return True
+        except ImportError as e:
+            return False
+
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Requires a CPMpy model as input, and will create the corresponding
+        Exact solver object xct_solver.
+
+        xct_solver can be modified externally before
+        calling solve(), a prime way to use more advanced solver features
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: None
+        """
+        if not self.supported():
+            raise Exception("Install the python 'exact' package to use this solver interface")
+
+        from exact import Exact as xct
+
+        assert subsolver is None, "Exact does not allow subsolvers."
+
+        # initialise the native solver object
+        self.xct_solver = xct()
+
+        # for solving with assumption variables,
+        self.assumption_dict = None
+
+        # objective can only be set once, so keep track of this
+        self.has_objective = False
+        self.objective_minimize = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="exact", cpm_model=cpm_model)
+
+    def getSolAndObj(self):
+        if not self.xct_solver.hasSolution():
+            self.objective_value_ = None
+            return False
+
+        # fill in variable values
+        exact_vals = self.xct_solver.getLastSolutionFor([self.solver_var(cpm_var) for cpm_var in self.user_vars])
+        i = 0
+        for cpm_var in self.user_vars:
```

Comment:
I think user_vars is a set, which is an unordered type.
I would not trust Python to iteration in this and the previous line in the same order...
maybe
```python
lst_vars = list(self.user_vars)
lst_vals = self.xct_solver.getLas...
for cpm_var, val in zip(lst_vars, lst_vals):
    cpm_var._value = val
    if....
```

Reply:
Relevant point, fixed.